File: lib/joker.rb
#
# Joker is a simple wildcard implementation that works much like Regexps.
#

#
# Implements wildcards for Ruby. Modeled after the Regexp class.
#
# This implementation supports the following special characters:
#
# - '?'   matches a single character
# - '*'   matches any number of characters, including 0
# - '\*'  matches a literal '*'
# - '\?'  matches a literal '?'
# - '\\'  matches a literal '\'
#
# NOTE that '\a' will match the literal string '\a'.
#
#   wild = Wildcard['Fairy?ake*']
#   wild =~ 'Fairycake'                     #=> true
#   wild =~ 'Fairyfakes are mean'           #=> true
#   wild =~ 'Fairysteakes are delicious'    #=> false
#
# Also there is a case sensitivity flag. By default it is set to true,
# but it can be turned off at construction time:
#
#   wild = Wildcard['Fairy?ake*', true]
#   wild =~ 'Fairycake'
#   wild =~ 'fairyCAKE'
#   wild =~ 'FaIrYfAkEs, the Movie'
#
# Furthermore, any given Wildcard expression must match the whole string:
#
#   wild = Wildcard['Fairy?ake']
#   wild =~ 'some Fairycake'                #=> false
#   wild =~ 'Fairycake is good for you'     #=> false
#
class Wildcard

    #
    # Boolean. Determines case sensitivity of the Wildcard.
    #
    # If this is true, the Wildcard will ignore case.
    #
    attr_reader :casefold

    #
    # The string from which the Wildcard was constructed.
    #
    attr_reader :source

    #
    # Creates a new Wildcard from the given string.
    # If casefold is true, the Wildcard will ignore case.
    #
    def initialize( wildcard_string, casefold = false )
        @source   = wildcard_string
        @casefold = casefold
        @regexp   =
            if casefold then Regexp.new(compile, Regexp::IGNORECASE)
            else             Regexp.new(compile)
            end
    end

    class << self

        #
        # Returns a new string with any characters escaped that would have
        # special meaning in a Wildcard.
        #
        def quote( string )
            string.gsub(%r{[\\?*]}) { |char| "\\#{char}" }
        end

        alias_method :[], :new
        alias_method :compile, :new
        alias_method :escape, :quote

    end

    def inspect
        %{Wildcard[#{@wild.inspect}]#{@casefold ? 'i' : ''}}
    end

    #
    # Matches the wildcard against $_:
    #
    #   $_ = 'I love fairycakes'
    #   ~Wildcard['*fairy*']        #=> true
    #
    def ~
        self =~ $_
    end

    #
    # Matches the Wildcard against the given string.
    #
    # NOTE: Since a wildcard has to match the whole string,
    # this method only returns true or false, not the position
    # of the match.
    #
    #   Wildcard['*fairy*'] =~ 'I love fairycake'   #=> true
    #   'I love fairycake' =~ Wildcard['*dairy*']   #=> false
    #
    def =~( string )
        !!(@regexp =~ string)
    end

    #
    # The case operator. Allows you to use Wildcards in case
    # expressions:
    #
    #   case 'I love fairycake'
    #   when Wildcard['*fairy*'] then puts 'fairy!'
    #   else puts 'no fairy...'
    #   end
    #
    def ===( object )
        !!(@regexp =~ object)
    end

    #
    # Compares to Wildcards for equality.
    #
    # Two wildcards are equal, if they were constructed from the
    # same string and have the same #casefold?().
    #
    def eql?( that )
        return false unless that.is_a?(Wildcard)
        @source == that.source && @casefold == that.casefold
    end

    alias_method :==, :eql?
    alias_method :casefold?, :casefold

    private

    #
    # Converts the wildcard string into a Regexp.
    # A simple parser, I just threw it in there, no
    # optimizations.
    #
    def compile
        ptr = 0
        compiled = '^'
        while ptr < @source.length
            snip = @source[ptr..ptr]
            case snip
            when '\\'
                lookahead = @source[ptr+1..ptr+1]
                case snip
                when '\\\\', '\\?', '\\*'
                    compiled << snip << lookahead
                else
                    compiled << Regexp.quote(lookahead)
                end
                ptr += 1
            when '?' then compiled << '.'
            when '*' then compiled << '.*'
            else          compiled << Regexp.quote(snip)
            end
            ptr += 1
        end
        compiled + '$'
    end

end

File: ext/joker/Joker.c
#include <malloc.h>
#include "Joker.h"
#include "Wildcard.h"
#include "compile.h"
#include "match.h"


void
Init_joker(void)  // {{{1
{
    class_Wildcard = rb_define_class("Wildcard", rb_cObject);
    rb_define_singleton_method(class_Wildcard, "new", class_method_new, -2);
    rb_define_method(class_Wildcard, "=~",  instance_operator_match, 1);
    rb_define_method(class_Wildcard, "===", instance_operator_case, 1);
}


/*
 * call-seq:
 *   wildcard =~ 'string' -> true or false
 *   'string' =~ wildcard -> true or false
 *
 * Matches the Wildcard against the given string.
 *
 * NOTE: Since a wildcard has to match the whole string,
 * this method only returns true or false, not the position
 * of the match.
 *
 *   Wildcard['*fairy*'] =~ 'I love fairycake'   #=> true
 *   'I love fairycake' =~ Wildcard['*dairy*']   #=> false
 *
 */
VALUE
instance_operator_match(self, string)  // {{{1
    VALUE  self;
    VALUE  string;
{
    Wildcard *    wildcard;
    const char *  cstring;
    long int      length;
    int           casefold;

    Data_Get_Struct(self, Wildcard, wildcard);
    cstring   = rb_str2cstr(string, &length);
    casefold  = RTEST(rb_iv_get(self, "@casefold"));

    if (Wildcard_match(wildcard, cstring, length, casefold)) {
        return Qtrue;
    } else {
        return Qfalse;
    }
}


/*
 * call-seq:
 *   wildcard === 'string' -> true or false
 *
 * The case operator. Allows you to use Wildcards in case
 * expressions:
 *
 *   case 'I love fairycake'
 *   when Wildcard['*fairy*'] then puts 'fairy!'
 *   else puts 'no fairy...'
 *   end
 *
 */
VALUE
instance_operator_case(self, object)  // {{{1
    VALUE  self;
    VALUE  object;
{
    Wildcard *    wildcard;
    const char *  cstring;
    long int      length;
    int           casefold;

    Check_Type(object, T_STRING);
    Data_Get_Struct(self, Wildcard, wildcard);
    cstring  = rb_str2cstr(object, &length);
    casefold  = RTEST(rb_iv_get(self, "@casefold"));

    if (Wildcard_match(wildcard, cstring, length, casefold)) {
        return Qtrue;
    } else {
        return Qfalse;
    }
}


/*
 * call-seq:
 *   Wildcard.new(wildcard_string, casefold = false) -> Wildcard
 *
 * Creates a new Wildcard from the given string.
 * If casefold is true, the Wildcard will ignore case.
 *
 * Raisess a SyntaxError if the given string could not
 * be interpreted as a Wildcard.
 *
 * Issues warnings to the console if the given Wildcard
 * was malformed.
 *
 * TODO: Wildcard.warn = false
 *
 */
VALUE
class_method_new(argc, argv, klass)  // {{{1
    int      argc;
    VALUE *  argv;
    VALUE    klass;
{
    VALUE         wildcard_string;
    VALUE         casefold;
    VALUE         new_object;
    Wildcard *    new_wildcard;
    const char *  wildcard_cstring;
    long int      string_length;

    rb_scan_args(argc, argv, "11", &wildcard_string, &casefold);
    if (NIL_P(casefold)) {
        casefold = Qfalse;
    }

    wildcard_cstring = rb_str2cstr(wildcard_string, &string_length);
    new_wildcard     = Wildcard_init();
    Wildcard_compile(wildcard_cstring, string_length, new_wildcard);
    new_object       = Data_Wrap_Struct(klass, 0, Wildcard_free, new_wildcard); 

    rb_iv_set(new_object, "@casefold", casefold);

    return new_object;
}

